File: rssp/cloud-sign/src/main/java/com/gov/crypto/cloudsign/config/SecurityConfig.java
package com.gov.crypto.cloudsign.config;

import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.annotation.web.configurers.AbstractHttpConfigurer;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.session.data.redis.config.annotation.web.http.EnableRedisHttpSession;

/**
 * Security configuration for cloud-sign service.
 * Uses Spring Session (Redis) for authentication - sessions are shared
 * with identity-service via Redis for SSO.
 */
@Configuration
@EnableWebSecurity
@EnableRedisHttpSession(redisNamespace = "crypto-session")
public class SecurityConfig {

    @Bean
    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
        return http
                .csrf(AbstractHttpConfigurer::disable)
                .authorizeHttpRequests(auth -> auth
                        .requestMatchers("/actuator/**").permitAll()
                        // All CSC endpoints need auth (Principal from session)
                        .requestMatchers("/csc/v1/**").authenticated()
                        .requestMatchers("/api/v1/credentials/**").authenticated()
                        .anyRequest().authenticated())
                // Session-based auth - Principal populated from Redis session
                .build();
    }
}

File: backend/api-gateway/src/main/kotlin/com/gov/crypto/apigateway/config/SecurityConfig.kt
package com.gov.crypto.apigateway.config

import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.config.Customizer
import org.springframework.security.config.annotation.web.reactive.EnableWebFluxSecurity
import org.springframework.security.config.web.server.ServerHttpSecurity
import org.springframework.security.web.server.SecurityWebFilterChain

@Configuration
@EnableWebFluxSecurity
class SecurityConfig {

    @Bean
    fun securityWebFilterChain(http: ServerHttpSecurity): SecurityWebFilterChain {
        return http
            .csrf { it.disable() }
            .authorizeExchange { exchanges ->
                exchanges
                    // Public endpoints - Auth (registration, login)
                    .pathMatchers("/api/v1/auth/**").permitAll()
                    // Public endpoints - PKI CA info (certificate download)
                    .pathMatchers("/api/v1/pki/ca/**").permitAll()
                    // Public endpoints - TSA (RFC 3161 timestamping)
                    .pathMatchers("/api/v1/tsa/**").permitAll()
                    // Require authentication for everything else
                    .anyExchange().authenticated()
            }
            .httpBasic(Customizer.withDefaults())
            .build()
    }
}
